Add a cluster diagnostic to check if master is also running as a node.

dgoodwin opened the pull request:
This is currently a requirement for Open vSwitch to ensure the master
is on the SDN and can communicate with other nodes.

Diagnostic uses the configured server URL in the kube config to perform a DNS
lookup for the IP(s) of the master. Care is taken to handle both IPv4 and IPv6
addresses being directly in the server URL already.

Once we've got a list of server IPs to check we scan the nodes via the
Kubernetes API, and check all of their reported IP addresses to look for a
match. No DNS lookup is done on the nodes, the IP addresses are already known.

If no node reports the IP of the master server, we warn the user that something
may be amiss and why.

NOTE: Technically speaking I have not actually tested the IPv6 bits outside of the unit tests. (if we even support such deployments today). If we can assume that Kube node addresses will contain strings formatted as golang's net.IP.String() would format them, there's a reasonable chance this will work. Worst case scenario is an inaccurate diagnostic warning.

dgoodwin commented:
Hit me with a [test] robots.

dgoodwin commented:
Editor supposedly configured to run gofmt on save, but somehow didn't get this issue, it shows up on CLI.

gofmt fix force pushed. Another [test] please.

brenton commented:
@dgoodwin, Would it be possible to check the value of the networkConfig.networkPluginName to skip this check if the `redhat/openshift-ovs-subnet` plugin or "the one for the multi-tenant ovs" isn't in use?  From a quick grep of the codebase it seems like `redhat/openshift-ovs-multitenant` is the name for the latter.

dgoodwin commented:
@brenton Hmm possibly, this is implemented as a cluster admin diagnostic so we can view all the node data. In theory it's possible that it is being run on a foreign host provides the credentials have cluster admin privs.

So the config file may not be on disk, however I could check to see if we can find a master config file, if so check the network plugin name and skip if necessary. If there's no file though we'd probably have to let it run I think.

Unless there's some way to access configuration via the API?

liggitt commented:
it is possible to provide the master config to diagnostics... if it is provided, can we use that info? There is no API for getting config info

dgoodwin commented:
Ah --master-config I see now, which implies --host. 

I will see if I can wire that up so it's passed through to the cluster admin diagnostics if provided, and also check for the default file location, and if found check the network plugin for the ovs names above and skip if we don't see a correct network plugin.

Is it acceptable to let it run if no config is available? (I'm thinking yes as a lot of folks might not be running this on the host or container itself)

brenton commented:
@dgoodwin, I would default to running the check even in cases where you don't have the master configuration file.  If you don't have access to the config file I think it would be OK to state in the warning message can be ignored if using a plugin other than the two Red Hat OVS related ones in their master config file.

@liggitt, if we have a kubeconfig with sufficient privileges is there a deterministic API call we could make to actually validate that the Master can proxy to pods correctly?  Would we simply search for a pod not running on the Master and then attempt to access it in some way?  That seems a little hokey and and would lean to non-deterministic warnings or error messages.  Another hack if we were really paranoid about the cluster network being misconfigured would be to attempt proxying to one of the "well-known" pod names like the router or the registry if that was safe.

What I'd prefer is to eventually have a network diagnostics pod that the diagnostic tool could optionally deploy to verify the network from within the cluster.  Then we could attempt accessing it from the master.  I think this would be something we tackle later though.

brenton commented:
That might be all we need to do.  I'll admit I'm a little out of touch with the problem.  @dgoodwin, if you're able to verify a working environment with that API call I would vote for adding it in.  We definitely want to make the warnings and error messages as useful as possible.

dgoodwin commented:
@liggit Can you clarify, is that proxy healthz check something from Kube or from Openshift? (guessing kube given the api vs oapi) And is that actually verifying that SDN connectivity is working between the master and the node?

This diagnostic is (currently) more limited in scope, purely checking that master is also running node, something that tripped me up and has reportedly surfaced with customers too.

There is a pre-existing diagnostic (NodeDefinitions) which analyzes the status of the nodes in the cluster. Right now it's just checking the Kube node status and informing the user if any are not ready. I am guessing that is not the same thing as checking this healthz info.

I kind of like the idea of fine grained separate diagnostics, so my proposal would be to implement a new separate diagnostic for healthz check, that looks really valuable if it verifies SDN is ok. It involves a ping and DNS lookups, which may slow down existing NodeDefinitions diagnostic quite a bit, so it feels like a good fit for a separate diagnostic to me.

Does the above sound good? 

Working on implementing the "skip MasterNode if master config is available and network plugin doesn't match" change now.

dgoodwin commented:
Updated with the "skip if master config is available and network plugin isn't OVS" change, and rebased off master.

Let me know thoughts on the above regarding healthz checks. (but I'm pitching separate PR)

brenton commented:
This looks good to me.  @liggitt, did you have anything you wanted to add?

As for a followup PR I could see us using a separate minimalistic connectivity health check on every invocation of the diagnostics tool to determine whether or not we should encourage the admin to run the more advanced SDN debug tooling that @danwinship demoed today.  Dan might even be able to point us to something more useful that could quickly determine if an admin needs to dig a little deeper using the SDN debug tool.

Dan, I'm not sure if you've had time to consider shipping the debug work you've done with the diagnostics cli but it might be worth considering so admins have it all in one place.

dgoodwin commented:
Re-[test] please.

danwinship commented:
> This is currently a requirement for Open vSwitch to ensure the master
> is on the SDN and can communicate with other nodes.

Hm? The master would only need to be on the SDN if it wanted to communicate directly with pods. It does not need to be on the SDN to communicate with the openshift-node services on the nodes.

The vagrant dev cluster setup does not set the master up as a node and everything works fine there...

danwinship commented:
> As for a followup PR I could see us using a separate minimalistic connectivity health check on every invocation of the diagnostics tool to determine whether or not we should encourage the admin to run the more advanced SDN debug tooling that @danwinship demoed today. Dan might even be able to point us to something more useful that could quickly determine if an admin needs to dig a little deeper using the SDN debug tool.
> 
> Dan, I'm not sure if you've had time to consider shipping the debug work you've done with the diagnostics cli but it might be worth considering so admins have it all in one place.

FYI the tool is hack/debug.sh in the openshift-sdn repo. At the moment, it's just a big dumb script that gathers every piece of possibly-useful information we could think of, for later analysis by a human being. Once 3.1 is out of the way we're planning to make it smarter so that it can diagnose at least some problems by itself. Until then, I don't think it's anything we want end users running on a regular basis.

brenton commented:
@danwinship, I agree you wouldn't expect the Master have to be part of the SDN but I remember it being required for users to be able to access jolokia running in the pods from the Web Console (which accessing the pod through the Master).

dgoodwin commented:
Not sure what that failing test is but I'd be quite shocked if it was anything in this PR (given the clean test run prior to the last change).

So is master must run node a safe assumption for most common deployments? FWIW this is a diagnostic warning (not error).

brenton commented:
Right now for OpenShift Enterprise we want customers to run the Master and the Node together to avoid support calls in cases like jolokia.  The idea behind this check was to help admins that may either be installing by hand or trying to creating their own scripts and we wanted to clue them in on a subtle feature their users are likely expecting that isn't obvious from an admin's perspective.  The default installer for OpenShift right now installs an unschedulable Node on each Master.

danwinship commented:
Maybe the long-term fix should be that we plug the master into the SDN even if it's not a node

brenton commented:
@danwinship, I've definitely heard that discussed.  This "requirement" was discovered very late in the initial 3.0 release cycle (days before GA) and this is just the current reality.

dgoodwin commented:
Rebase on master, and softened the verbage a bit:

WARN:  [DClu3004 from diagnostic MasterNode@openshift/origin/pkg/diagnostics/cluster/master_node.go:176]
       Unable to find a node matching the cluster server IP.
       This may indicate the master is not also running a node, and is unable
       to proxy to pods over the Open vSwitch SDN.

dgoodwin commented:
Updated to fix the (new?) govet issues.

deads2k commented:
jenkins issue re[test]

dgoodwin commented:
Test failures do not appear related as far as I can tell, suspect this PR may be ready to roll.

dgoodwin commented:
Rebased on master, another [test] please.

Should be ready for merge afaik.

danmcp commented:
@brenton Review and merge please.

brenton commented:
Re-reading all the comments I'm going to submit this for merging because I think it will help reduce support calls.  I think the warning is fairly clear that it only applies to users of the openvswitch plugins.  We'll definitely reassess the situation if customers are still getting confused. [merge]
